File: 40. 组合总和 II/Solution.java
class Solution {
    LinkedList<Integer> path = new LinkedList<>();
    List<List<Integer>> ans = new ArrayList<>();
    int sum = 0;
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        backTracking(candidates, target, 0);
        return ans;
    }

    private void backTracking(int[] candidates, int target, int startIndex){
        if(sum == target){
            ans.add(new ArrayList(path));
        }
        for(int i = startIndex; i < candidates.length; i++){
            if(sum + candidates[i] > target){
                break;
            }
            if(i > startIndex && candidates[i] == candidates[i-1]){
                continue;
            }
            sum += candidates[i];
            path.add(candidates[i]);
            backTracking(candidates, target, i+1);
            sum -= candidates[i];
            path.removeLast();
        }
    }
}

File: 40. 组合总和 II/Solution.py
class Solution:
    def traversal(self, ans, candidates, target, path, Sum, start):
        if Sum == target:
            ans.append(path[:])
            return
        for i in range(start, len(candidates)):
            if i > start and candidates[i] == candidates[i-1]: # 去除重复的组合。说明前一个树枝，使用了candidates[i - 1]，也就是说同一树层使用过candidates[i - 1]。
                continue
            if Sum + candidates[i] > target: # 剪枝
                break
            Sum += candidates[i]
            path.append(candidates[i])
            self.traversal(ans, candidates, target, path, Sum, i+1)
            Sum -= candidates[i]
            path.pop()

    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        ans = []
        path = []
        self.traversal(ans, candidates, target, path, 0, 0)
        return ans
